ryan_savill/principal_feature_analysis/readme.md

```diff
@@ -0,0 +1,902 @@
+                    q = i
+                    break
+                    
+        A_q = pca.components_.T[:,:q]
+        
+        clusternumber = min([q + self.diff_n_features, X.shape[1]])
+        
+        kmeans = KMeans(n_clusters= clusternumber).fit(A_q)
+        clusters = kmeans.predict(A_q)
+        cluster_centers = kmeans.cluster_centers_
+
+        dists = defaultdict(list)
+        for i, c in enumerate(clusters):
+            dist = euclidean_distances([A_q[i, :]], [cluster_centers[c, :]])[0][0]
+            dists[c].append((i, dist))
+
+        self.indices_ = [sorted(f, key=lambda x: x[1])[0][0] for f in dists.values()]
+        self.features_ = X[:, self.indices_]
+        
+    def fit_transform(self,X):    
+        sc = StandardScaler()
+        X = sc.fit_transform(X)
+
+        pca = PCA().fit(X)
+        
+        if not self.q:
+            explained_variance = pca.explained_variance_ratio_
+            cumulative_expl_var = [sum(explained_variance[:i+1]) for i in range(len(explained_variance))]
+            for i,j in enumerate(cumulative_expl_var):
+                if j >= self.explained_var:
+                    q = i
+                    break
+                    
+        A_q = pca.components_.T[:,:q]
+        
+        clusternumber = min([q + self.diff_n_features, X.shape[1]])
+        
+        kmeans = KMeans(n_clusters= clusternumber).fit(A_q)
+        clusters = kmeans.predict(A_q)
+        cluster_centers = kmeans.cluster_centers_
+
+        dists = defaultdict(list)
+        for i, c in enumerate(clusters):
+            dist = euclidean_distances([A_q[i, :]], [cluster_centers[c, :]])[0][0]
+            dists[c].append((i, dist))
+
+        self.indices_ = [sorted(f, key=lambda x: x[1])[0][0] for f in dists.values()]
+        self.features_ = X[:, self.indices_]
+        
+        return X[:, self.indices_]
+    
+    def transform(self, X):
+        return X[:, self.indices_]
+
+# Testing
+pfa = PFA(diff_n_features=1, explained_var= 0.95)
+pfa.fit_transform(regprops)
+
+featurekeys = [regprops.keys().tolist()[i] for i in pfa.indices_]
+featurekeys
+```
+
+
+
+
+    ['area',
+     'centroid-0',
+     'centroid-1',
+     'solidity',
+     'mean_intensity',
+     'image_stdev',
+     'touching neighbor count']
+
+
+
+This is a modification of the provided code in that it tries to force you to choose a subset. If you don't manually say what the subset is, it will get a subset that explains 95% of the variance or the amount of variance that you determine. As described in the paper the number of clusters (and with that the number of features) should be larger than the subset of eigenvectors chosen. This is implemented by not determining the number of features but only setting how many more features than the subset should be chosen. Apart from this a transform function was added as well as a fit_transform function, similarly to other scikit learn functions. (DISCLAIMER FOR CODE USAGE?)
```

Comment:
```suggestion
This is a modification of the provided code in that it tries to force you to choose a subset. If you don't manually say what the subset is, it will get a subset that explains 95% of the variance or the amount of variance that you determine. As described in the paper, the number of clusters (and with that the number of features) should be larger than the subset of eigenvectors chosen. This is implemented by not determining the number of features but only setting how many more features than the subset should be chosen. Apart from this a transform function was added as well as a fit_transform function, similarly to other scikit learn functions.